pkg/controllers/nodeclass/status/ami/controller.go

```diff
@@ -0,0 +1,123 @@
+*/
+
+package ami
+
+import (
+	"context"
+	"github.com/aws/karpenter-provider-aws/pkg/apis/v1beta1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/amifamily"
+	"github.com/awslabs/operatorpkg/reasonable"
+	"github.com/awslabs/operatorpkg/singleton"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	corev1beta1 "sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+	"sort"
+	"time"
+)
+
+type Controller struct {
+	kubeClient  client.Client
+	amiProvider amifamily.Provider
+}
+
+// NewController is a constructor
+func NewController(kubeClient client.Client, amiProvider amifamily.Provider) *Controller {
+
+	return &Controller{
+		kubeClient:  kubeClient,
+		amiProvider: amiProvider,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "nodeclass.ami")
+
+	if !controllerutil.ContainsFinalizer(nodeClass, v1beta1.TerminationFinalizer) {
+		stored := nodeClass.DeepCopy()
+		controllerutil.AddFinalizer(nodeClass, v1beta1.TerminationFinalizer)
+		if err := c.kubeClient.Patch(ctx, nodeClass, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	stored := nodeClass.DeepCopy()
+	amis, err := c.amiProvider.List(ctx, nodeClass)
+	var errs error
+	if err != nil {
+		nodeClass.StatusConditions().SetFalse(v1beta1.ConditionTypeAMIsReady, "UnresolvedAMIs", "Unable to resolve AMI")
+		errs = multierr.Append(errs, err)
+	} else {
+		if len(amis) == 0 {
+			nodeClass.Status.AMIs = nil
+			nodeClass.StatusConditions().SetFalse(v1beta1.ConditionTypeAMIsReady, "UnresolvedAMIs", "Unable to resolve AMI")
+		} else {
+			nodeClass.Status.AMIs = lo.Map(amis, func(ami amifamily.AMI, _ int) v1beta1.AMI {
+				reqs := lo.Map(ami.Requirements.NodeSelectorRequirements(), func(item corev1beta1.NodeSelectorRequirementWithMinValues, _ int) v1.NodeSelectorRequirement {
+					return item.NodeSelectorRequirement
+				})
+
+				sort.Slice(reqs, func(i, j int) bool {
+					if len(reqs[i].Key) != len(reqs[j].Key) {
+						return len(reqs[i].Key) < len(reqs[j].Key)
+					}
+					return reqs[i].Key < reqs[j].Key
+				})
+				return v1beta1.AMI{
+					Name:         ami.Name,
+					ID:           ami.AmiID,
+					Requirements: reqs,
+				}
+			})
+			nodeClass.StatusConditions().SetTrue(v1beta1.ConditionTypeAMIsReady)
+		}
+	}
+	if !equality.Semantic.DeepEqual(stored, nodeClass) {
+		if err = c.kubeClient.Status().Update(ctx, nodeClass); err != nil {
```

Comment:
Why does this need to be an update over a patch? We're only updating the existing status condition right? Do we expect any other controllers to be updating this status condition which we could conflict with? Same comment for all other controllers.

Reply:
I tried to follow what we have done with most of our controllers for status updates. Is there a benefit of using patch over update?